In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
import mysql.connector as connector
import os
from dotenv import load_dotenv

In [4]:
# Load environment variables from .env file

load_dotenv()

True

<h3 style="font-size: 18px; text-align: center"><b>Database Connection and Retrive Data from Database</b></h3>

In [5]:
# Fetch database connection parameters from environment variables
db_params = {
    'dbname': os.environ.get('DB_NAME'),
    'table': os.environ.get('DB_TABLE_NAME'),
    'host': os.environ.get('DB_HOST'),
    'port': os.environ.get('DB_PORT'),
    'userame': os.environ.get('DB_USERNAME'),
    'password': os.environ.get('DB_PASSWORD'),
}

In [6]:
connection = connector.connect(host=db_params['host'], user=db_params['userame'], password=db_params['password'], database=db_params['dbname'])

In [7]:
# List of Databases

cursor = connection.cursor()
cursor.execute('show databases')
for database in cursor:
    print(database)

('information_schema',)
('project_sales',)


In [8]:
# List of Tables

db_tables = pd.read_sql_query('show tables', connection)
print(db_tables)

  Tables_in_project_sales
0                    data


In [9]:
query = 'select * from data'
output_from_database = pd.read_sql_query(query, connection)
print(output_from_database.shape, "Rows as Columns")

(7422, 9) Rows as Columns


In [10]:
# saving the dataset to csv for the local usage.
# Since The data in the db should not be changed or updated or deleted or added any new data.

# Define the directory and file name
directory = 'data'
if not os.path.exists(directory):
    os.makedirs(directory)

data_path = os.path.join(directory, 'data.csv')
output_from_database.to_csv(data_path, index=False)

print(f"Data saved to {data_path}")

Data saved to data\data.csv


In [11]:
# Load the dataset from local
data = pd.read_csv('data/data.csv')

<h3 style="font-size: 18px; text-align: center;"><b>Basic Checks</b></h3>

In [12]:
# First 5 datas from the dataset
data.head(5)

Created  Product_ID   Source      Mobile               EMAIL  \
0  14-11-2018 10:05         NaN  Website  984XXXXXXX  aXXXXXXX@gmail.com   
1  14-11-2018 09:22         NaN  Website     XXXXXXX             #VALUE!   
2  14-11-2018 09:21         NaN  Website     XXXXXXX  dXXXXXXX@yahoo.com   
3  14-11-2018 08:46         NaN  Website     XXXXXXX  wXXXXXXX@gmail.com   
4  14-11-2018 07:34         NaN  Website     XXXXXXX  cXXXXXXX@gmail.com   

      Sales_Agent Location Delivery_Mode Status  
0  Sales-Agent-11      NaN        Mode-5   Open  
1  Sales-Agent-10      NaN        Mode-5   Open  
2  Sales-Agent-10      NaN        Mode-5   Open  
3  Sales-Agent-10      NaN        Mode-5   Open  
4  Sales-Agent-10      NaN        Mode-5   Open

In [13]:
# Last 5 datas from the dataset
data.tail(5)

Created  Product_ID            Source Mobile  \
7417  28-04-2018 09:45         9.0              Call    NaN   
7418  28-04-2018 09:43        15.0              Call    NaN   
7419  28-04-2018 09:20         5.0  Live Chat-Direct    NaN   
7420  28-04-2018 08:04        21.0          CRM form    NaN   
7421  28-04-2018 07:54        25.0           Website    NaN   

                   EMAIL     Sales_Agent         Location Delivery_Mode  \
7417  aXXXXXXX@gmail.com   Sales-Agent-6           Mumbai        Mode-4   
7418             #VALUE!  Sales-Agent-12  Other Locations        Mode-5   
7419  sXXXXXXX@gmail.com  Sales-Agent-11        Bangalore        Mode-1   
7420  YXXXXXXX@gmail.com   Sales-Agent-4  Other Locations        Mode-1   
7421  cXXXXXXX@gmail.com   Sales-Agent-3          Chennai        Mode-1   

              Status  
7417            LOST  
7418            LOST  
7419  Not Responding  
7420    Just Enquiry  
7421       CONVERTED

In [14]:
# Total No.of Rows and Columns in dataset

rows, col = data.shape
print ("Dimensions of dataset: {}" . format (data.shape))
print ('Rows:', rows, '\nColumns:', col)

Dimensions of dataset: (7422, 9)
Rows: 7422 
Columns: 9


In [15]:
# Columns in dataset

columns = data.columns
for i in range(len(columns)):
    print(f'Column {i}: {columns[i]}')

Column 0: Created
Column 1: Product_ID
Column 2: Source
Column 3: Mobile
Column 4: EMAIL
Column 5: Sales_Agent
Column 6: Location
Column 7: Delivery_Mode
Column 8: Status


In [16]:
# Information about Columns in dataset

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7422 entries, 0 to 7421
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Created        7422 non-null   object 
 1   Product_ID     7364 non-null   float64
 2   Source         7405 non-null   object 
 3   Mobile         5612 non-null   object 
 4   EMAIL          7422 non-null   object 
 5   Sales_Agent    7399 non-null   object 
 6   Location       7364 non-null   object 
 7   Delivery_Mode  7422 non-null   object 
 8   Status         7422 non-null   object 
dtypes: float64(1), object(8)
memory usage: 522.0+ KB


<h3 style="font-size: 18px;"><b>Insights</b></h3>

- <span style="font-size: 14px;">There are totaly <b>7422</b> Entries</span>

    - <span style="font-size: 14px;">The <b>Created</b>, <b>Email</b>, <b>Delivery_Mode</b> and <b>Status</b> has <b>No</b> Null values.</span>
    
    - <span style="font-size: 14px;">The Product_ID has <b>58</b> has Null values.</span>

    - <span style="font-size: 14px;">The Source has <b>17</b> has Null values.</span>

    - <span style="font-size: 14px;">The Mobile has <b>1810</b> has Null values.</span>

    - <span style="font-size: 14px;">The Sales_Agent has <b>23</b> has Null values.</span>

    - <span style="font-size: 14px;">The Location has <b>58</b> has Null values.</span>



In [17]:
# Statistical analysis of the dataset

data.describe().T

count       mean       std  min   25%   50%   75%   max
Product_ID  7364.0  15.947311  6.072937  0.0  12.0  18.0  19.0  28.0

In [18]:
# Droping the Columns like Mobile, created and Email since they are not needed for the Analysis.

data.drop(['Mobile', 'Created', 'EMAIL'], axis=1, inplace=True)

In [19]:
# data after dropping the unwanted Columns.
data.head(5)

Product_ID   Source     Sales_Agent Location Delivery_Mode Status
0         NaN  Website  Sales-Agent-11      NaN        Mode-5   Open
1         NaN  Website  Sales-Agent-10      NaN        Mode-5   Open
2         NaN  Website  Sales-Agent-10      NaN        Mode-5   Open
3         NaN  Website  Sales-Agent-10      NaN        Mode-5   Open
4         NaN  Website  Sales-Agent-10      NaN        Mode-5   Open

In [20]:
# Replacing the most frequent column value in each column.

for column in data.select_dtypes(include='object').columns:
    most_frequent_column = data[column].mode()[0]
    data[column].fillna(most_frequent_column, inplace=True)

In [21]:
# Data after Replacing the most frequent column value in each column.
data.head(5)

Product_ID   Source     Sales_Agent         Location Delivery_Mode Status
0         NaN  Website  Sales-Agent-11  Other Locations        Mode-5   Open
1         NaN  Website  Sales-Agent-10  Other Locations        Mode-5   Open
2         NaN  Website  Sales-Agent-10  Other Locations        Mode-5   Open
3         NaN  Website  Sales-Agent-10  Other Locations        Mode-5   Open
4         NaN  Website  Sales-Agent-10  Other Locations        Mode-5   Open

In [22]:
most_frequent_Product_ID = data['Product_ID'].mode()[0]

# Replace the frequent value for the Product_ID column
data['Product_ID'].fillna(most_frequent_Product_ID, inplace=True)

# Replaced data
data.head(5)

Product_ID   Source     Sales_Agent         Location Delivery_Mode Status
0        18.0  Website  Sales-Agent-11  Other Locations        Mode-5   Open
1        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Open
2        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Open
3        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Open
4        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Open

- The above steps implies that all the rows in the dataset are replaced with the most frequent column value in each column.

In [23]:
data.to_csv('data/new_data.csv', index=False)

In [24]:
updated_data = pd.read_csv('data/new_data.csv')

In [25]:
updated_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7422 entries, 0 to 7421
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Product_ID     7422 non-null   float64
 1   Source         7422 non-null   object 
 2   Sales_Agent    7422 non-null   object 
 3   Location       7422 non-null   object 
 4   Delivery_Mode  7422 non-null   object 
 5   Status         7422 non-null   object 
dtypes: float64(1), object(5)
memory usage: 348.0+ KB


In [26]:
# Compressing and Merging the labes in column

updated_data.Source.replace(['Live Chat-Direct','Live Chat-Google Organic','Live Chat -PPC','Live Chat-Blog','Live Chat-Quora',
                    'Live Chat-CPC','Live Chat-Google Ads','Live Chat-Adwords Remarketing','Live Chat-Youtube',
                    'Live Chat-Justdial'],'Live Chat',inplace=True)
updated_data.Source.replace(['Existing Client','CRM form','Personal Contact'],'Existing Customer',inplace=True)
updated_data.Source.replace('By Recommendation','Customer Referral',inplace=True)
updated_data.Source.replace(['US Website','Just Dial'],'Website',inplace=True)
updated_data.Source.replace(['E-mail Campaign','SMS Campaign','E-Mail Message','Other'],'Campaign',inplace=True)

In [27]:
updated_data.Location.replace(['UAE','USA','UK','AUSTRALIA','Singapore','Malaysia','EUROPE'],'Foreign',inplace=True)
updated_data.Location.replace(['Howrah','Kolkata','Trivandrum'],'Other Locations',inplace=True)

In [28]:
# Defining 'Status' as 'Good' or 'Bad'
updated_data.Status.replace(['CONVERTED','converted','In Progress Positive','Potential','Long Term','Open'],'Good',inplace=True)
updated_data.Status.replace(['LOST','In Progress Negative','Not Responding','Junk Lead','Just Enquiry'],'Bad',inplace=True)

In [29]:
updated_data.head(5)

Product_ID   Source     Sales_Agent         Location Delivery_Mode Status
0        18.0  Website  Sales-Agent-11  Other Locations        Mode-5   Good
1        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Good
2        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Good
3        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Good
4        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Good

In [30]:
updated_data.to_csv('data/new_data.csv', index=False)

In [32]:
updated_data.head(5)

Product_ID   Source     Sales_Agent         Location Delivery_Mode Status
0        18.0  Website  Sales-Agent-11  Other Locations        Mode-5   Good
1        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Good
2        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Good
3        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Good
4        18.0  Website  Sales-Agent-10  Other Locations        Mode-5   Good

In [33]:
# Check for missing values in datasets.

updated_data.isnull().sum()

Product_ID       0
Source           0
Sales_Agent      0
Location         0
Delivery_Mode    0
Status           0
dtype: int64

- The above check implies that there is no null value in dataset.

In [34]:
# check for duplicates in datasets.

print(f'The Number of duplicates in the given dataset is {updated_data.duplicated().sum()}')

The Number of duplicates in the given dataset is 4511


In [37]:
# Identify all the duplicate rows
duplicates = updated_data[updated_data.duplicated()]
duplicates

Product_ID     Source     Sales_Agent         Location Delivery_Mode  \
2           18.0    Website  Sales-Agent-10  Other Locations        Mode-5   
3           18.0    Website  Sales-Agent-10  Other Locations        Mode-5   
4           18.0    Website  Sales-Agent-10  Other Locations        Mode-5   
5           18.0    Website  Sales-Agent-10  Other Locations        Mode-5   
6           18.0    Website  Sales-Agent-10  Other Locations        Mode-5   
...          ...        ...             ...              ...           ...   
7415        21.0  Live Chat   Sales-Agent-4        Bangalore        Mode-1   
7416        19.0  Live Chat   Sales-Agent-9  Other Locations        Mode-3   
7418        15.0       Call  Sales-Agent-12  Other Locations        Mode-5   
7419         5.0  Live Chat  Sales-Agent-11        Bangalore        Mode-1   
7421        25.0    Website   Sales-Agent-3          Chennai        Mode-1   

     Status  
2      Good  
3      Good  
4      Good  
5      Good  
6      Good  
...     ...  
7415    Bad  
7416    Bad  
7418    Bad  
7419    Bad  
7421   Good  

[4511 rows x 6 columns]

In [40]:
# check number of duplicates is matching with the duplicates variable in dataset.
print(f"Are there 4511 duplicates in dataset? {len(duplicates) == updated_data.duplicated().sum()}")

Are there 4511 duplicates in dataset? True


In [41]:
# no of unique elements in each columns

updated_data.nunique().to_frame().T

Product_ID  Source  Sales_Agent  Location  Delivery_Mode  Status
0          29       6           12         8              5       2

<h3 style="font-size: 18px; text-align: center;"><b>Data Preprocessing</b></h3>